### PR TITLE
Add Guice dependency and inject ExtensionsRunner into Extension

### PR DESCRIPTION
**Pull request by dbwiddis**
Signed-off-by: Daniel Widdis <widdis@gmail.com>

### Description

Use Guice for dependency injection.  

First of two or more PRs, splitting up into logical chunks to speed review.  

This PR just adds the dependency and initial injection setup.  Future PRs:
 - will integrate with `getActions()` implementation (pending merge of #434) to improve action handling
 - will add additional components necessary for AD extension 

This PR:
 - Injects (via lambda) a Guice `Module` which binds classes and makes them available via `@Inject`. 
 - Binds other objects associated with `createComponents()` (more to come in the future)
 - Injects the `ExtensionsRunner` into the `BaseExtension` as a sample/test case
 - Makes the `createComponents()` method on the `Extension` interface default, removing its arguments and removing the requirement for `BaseExtension` to mange/provide these objects (as they will be injected and available directly)

### Issues Resolved

Part of #368 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by joshpalis — APPROVED** [review]
This looks great, thanks for implementing Guice within the project, this will definitely simply integration later down the line.

**Review by ryanbogan — DISMISSED** [review]

**Review by joshpalis — APPROVED** [review]

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 1]
```diff
@@ -154,10 +164,34 @@ protected ExtensionsRunner(Extension extension) throws IOException {
         // save custom transport actions
         this.transportActions = new TransportActions(extension.getActionsMap());
 
-        ThreadPool threadPool = new ThreadPool(this.getSettings());
+        Guice.createInjector(b -> {
+            b.bind(ExtensionsRunner.class).toInstance(this);
+            b.bind(Extension.class).toInstance(extension);
+
+            b.bind(NamedXContentRegistry.class).toInstance(getNamedXContentRegistry().getRegistry());
+            b.bind(ThreadPool.class).toInstance(getThreadPool());
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:172]: Need to understand this. Do we need ThreadPool in SDK?

Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:172]: This is SDK's own thread pool.  

Do we need it?  No.  Do we want a single common thread pool that the whole extension can share if it so chooses, to balance its resources?  Why not?

Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:172]: Also from a PR standpoint, this does not change the previous behavior of a threadpool being created (line 157) and passed to the extension via create components.  Just changes the syntax of how an extension accesses that object.

Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:172]: Makes sense. Do we need to bind ThreadPool then, as it is specific to the SDK?

**Review by ryanbogan — APPROVED** [review]

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 1 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:172]: In theory we might make this object available for in-proc extensions to be passed as a param.  I don't know, but we can always easily remove it before GA.

**Review by owaiskazi19 — APPROVED** [review]

**Pull request merged**